### PR TITLE
intercept-build to work with response files.

### PR DESCRIPTION
**Pull request by rasjani**
More info: https://msdn.microsoft.com/en-us/library/3te4xt0y.aspx

QT's qmake nmake generator generates makefile that puts all the
compilation units into response file. Thus, when compiling with
intercept-build to generate compilation_database.json, it won't get
populated due to _split_command not being able to detect any source
files in the intercepted call.

This patch will read the response file and append its data as
commandline parameters.

Includes unittest - essentially it covers also the functionality
so i dont see any need for "proper" functional test.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,4 @@
 .idea
```
Comment (rizsotto) [inline .gitignore]: Could you cut this out from the PR? This is not related to this topic!

File: libscanbuild/intercept.py [hunk 2]
```diff
@@ -163,6 +163,15 @@ def write_exec_trace(filename, entry):
     :param filename:    path to the output execution trace file,
     :param entry:       the Execution object to append to that file. """
 
+    for param in entry.cmd[:]:
```
Comment (rizsotto) [inline libscanbuild/intercept.py]: Can this extracted to a separate method? Which would take the commands and would return a fixed command line. Would like to test it separately if we can.

File: tests/unit/test_intercept.py [hunk 3]
```diff
@@ -30,6 +30,29 @@ def test_read_write_exec_trace(self):
...
+            response_file_two = os.path.join(tmp_dir, 'response2.jom')
+            input_one = Execution(
+                pid=123,
+                cwd='/path/to/here',
+                cmd=['cc', '-c', '@'+response_file_one, 'this.c',
+                     '@'+response_file_two])
+            output_one = Execution(
+                pid=123,
+                cwd='/path/to/here',
+                cmd=['cc', '-c', 'that.c', 'this.c', '-Idoes_not_exists'])
```
Comment (rizsotto) [inline tests/unit/test_intercept.py]: I don't think it's a valid compiler call. You can't specify `-c` and two source files at the same time. (I know it's just a bunch of string in a text file, but tests are examples which helps people to understand what's going on.)

File: .gitignore (hunk 1 above)
Comment (rasjani) [inline .gitignore]: removed..

File: libscanbuild/intercept.py (hunk 2 above)
Comment (rasjani) [inline libscanbuild/intercept.py]: and extracted into a separate method with separate test case ..

File: tests/unit/test_intercept.py (hunk 3 above)
Comment (rasjani) [inline tests/unit/test_intercept.py]: im not hugely into memorizing what compiler calls do work and what compiler .. but one can definitely compile multiple units  like that with cl/clang-cl - thats quite what the whole response file is usually doing: it containts multiple source entries .. 

But yeah, i made the test data to reflect your point of view so that users wont get confused.

File: tests/unit/test_intercept.py [hunk 4]
```diff
@@ -30,6 +30,41 @@ def test_read_write_exec_trace(self):
             result = sut.parse_exec_trace(temp_file)
             self.assertEqual(input_one, result)
 
+    def test_expand_cmd_with_response_files(self):
```
Comment (rizsotto) [inline tests/unit/test_intercept.py]: Tests looks good to me now. Thanks for fixing it!

File: libscanbuild/intercept.py [hunk 5]
```diff
@@ -154,6 +154,19 @@ def intercept_compiler_wrapper(_, execution):
         logging.warning(message_prefix, 'io problem')
 
 
+def expand_cmd_with_response_files(cmd):
```
Comment (rizsotto) [inline libscanbuild/intercept.py]: Could you write pydoc comment for the method? A short one liner about that this method transform the command line arguments. Then a bit explanation, when it is happening (please mention windows, qmake with nmake as you did in the PR comment). And a short comments about the argument and return value. Please mention that this code might throw exception too.

File: libscanbuild/intercept.py [hunk 6]
```diff
@@ -154,6 +154,19 @@ def intercept_compiler_wrapper(_, execution):
...
+    for param in cmd[:]:
+        if param[:1] == '@' and os.path.isfile(param[1:]):
+            # can throw but intercap_compiler_wrapper should catch
+            with open(param[1:], 'r') as f:
```
Comment (rizsotto) [inline libscanbuild/intercept.py]: Could you extract this into a method too? :) What I see here is that reading the file and change the list element at the same place. Maybe to read the file and return the stripped content would deserve a separate method. What do you think? It could also placeholder to a comment which describes the file format of these files.

File: libscanbuild/intercept.py [hunk 7]
```diff
@@ -154,6 +154,19 @@ def intercept_compiler_wrapper(_, execution):
...
+                input_buffer = f.read()
+                idx = cmd.index(param)
+                del cmd[idx]
```
Comment (rizsotto) [inline libscanbuild/intercept.py]: Can we do the transformation without the deletion-insertion? I was thinking about a list comprehension solution like `[from_respone_file(arg) for arg in args if is_response_file(arg)]`. Where the `from_response_file` is the method which takes '@filename' and return the content (see previous comment). And the `is_response_file` checks the `@` prefix and the file existence. These methods can declared inside the method, but documented.

File: libscanbuild/intercept.py [hunk 8]
```diff
@@ -154,6 +154,36 @@ def intercept_compiler_wrapper(_, execution):
...
+        """ checks if the given command line argument is response file  """
+        return param[:1] == '@' and os.path.isfile(param[1:])
+
+    def from_response_file(filename):
+        """ read and return command line argument list from response file
+
+        can throw IOException but typically can as for now, its caught in
+        `intercept_compiler_wrapper` level.  """
+        with open(filename[1:], 'r') as f:
+            return [arg.strip() for arg in f.read().split()]
```
Comment (rasjani) [inline libscanbuild/intercept.py]: At the moment (as off qt 5.6.2) does not add compiler flags into a response file - only compilation unit(s). There's spaces in front and a line feed at the end. One compilation per line. This approach is visible if you check the generated makefile - you can also probably see the generated syntax if you just pass appropriate spec parameter to qmake even if you are not on windows.  

Also, if I remember correctly, ms documentation points out that response file should be single parameter per line.  And also coverity does honour this format but yeah.

**Comment by rasjani** [comment]
Damn no edit.  Yeah, checked the docs and I remembered the spec incorrectly. However, qmake based makefiles do only put sources in response files but handwritten or by other tools, single line could contain other approaches. 

Will update the pr accordingly.

File: libscanbuild/intercept.py (hunk 8 above)
Comment (rasjani) [inline libscanbuild/intercept.py]: http://daviddeley.com/autohotkey/parameters/parameters.htm#WINCRULES

Comment (rizsotto) [inline libscanbuild/intercept.py]: Sweet! It looks `shell_split` will deal with it smooth, but I will write some tests against it. Thanks!

File: libscanbuild/intercept.py [hunk 9]
```diff
@@ -154,6 +154,36 @@ def intercept_compiler_wrapper(_, execution):
...
+    by reading options from so called temporary "response" files. These files
+    are ascii encoded and can contain compiler and linker flags and/or
+    compilation units.
+
+    For example, QT's qmake generates nmake based makefiles where the response
+    file contains all compilation units. """
+
+    def is_response_file(param):
+        """ checks if the given command line argument is response file  """
+        return param[:1] == '@' and os.path.isfile(param[1:])
```
Comment (rasjani) [inline libscanbuild/intercept.py]: pydocs capitalized and punctuations - done
param[:1] to param[0] - done

File: libscanbuild/intercept.py [hunk 10]
```diff
@@ -154,6 +154,36 @@ def intercept_compiler_wrapper(_, execution):
...
+
+    def from_response_file(filename):
+        """ read and return command line argument list from response file
+
+        can throw IOException but typically can as for now, its caught in
```
Comment (rasjani) [inline libscanbuild/intercept.py]: done

File: libscanbuild/intercept.py (hunk 8 above)
Comment (rasjani) [inline libscanbuild/intercept.py]: shell_split used now .. Did you ever write more tests against it ? I have a recollection that i've tried to use it while back and noticed some issues in windows..  Maybe it would be good time now to extend those tests ?

**Comment by rizsotto** [comment]
Thanks to come back with it. Will check it this week.

**Comment by rizsotto** [comment]
It's merged to master. Please check it out if it works for you!

Also if you have some idea how to run the existing `qmake` functional test on windows, please share it. I would really appreciate it... I was [struggle](https://ci.appveyor.com/project/rizsotto/scan-build/build/2.0.220/job/h8eyk1x0b6d0lo7h) to link the binary against the library.
